src/hotspot/share/runtime/deoptimization.cpp

```diff
@@ -2494,7 +2495,10 @@ Deoptimization::query_update_method_data(MethodData* trap_mdo,
     // Find the profile data for this BCI.  If there isn't one,
     // try to allocate one from the MDO's set of spares.
     // This will let us detect a repeated trap at this point.
-    pdata = trap_mdo->allocate_bci_to_data(trap_bci, reason_is_speculate(reason) ? compiled_method : nullptr);
+    {
+      MutexLocker ml(trap_mdo->extra_data_lock());
```

Comment:
Doesn't the lock have to be held over the lifetime of the pdata variable?  Otherwise an intervening safepoint could repack the MDO rendering pdata invalid.

Reply:
Good point. That was done in a similar place above but not here.

Reply:
Hmm. So is it that I just have to have a lock on `allocate_bci_to_data`, or do I have to protect `pdata`? Because `pdata` is returned from `Deoptimization::query_update_method_data`, and so I would have to make the lock take a wider scope than that function, right?

Reply:
The lock needs to be held while accessing the contents of pdata which is very complicated in this particular usage pattern.  Maybe the caller needs to refetch the pointer under a lock instead of passing it out of this method.

Reply:
I see.
And how about other usages of `ProfileData` from the `mdo`, like these (there is a few variants of them around):
```
ProfileData* data = mdo->first_data();
ProfileData* data = mdo->bci_to_data(bci);
```

Reply:
And there are 2 uses of `query_update_method_data`. One does not use the return `pdata`. The other uses it and in some cases updates it. Do you think it is safe to just re-fetch it, or would that potentially cut some connection between the two that should not be cut?
The alternative is just to already get the lock before calling `query_update_method_data`.

Reply:
I think that anything that can return data from the extra data section is a potential danger.  bci_to_data calls bci_to_extra_data at the end so it seems potentially unsafe which seems like a huge problem since that's used all over the place.  Whether the callers are actually getting or expecting record from extra data is unclear.  I would suspect that most places where it's used there should already be a preallocated record.  The concurrent repacking really makes it hard to ensure the accesses are safe.  I think the API would need to make a stronger split between preallocated records and records which might come from the extra data section.  I'm honestly not sure how to make this truly safe.

Reply:
Ok. So this issue is much bigger than `query_update_method_data` and `allocate_bci_to_data`, is what you are saying. Sounds like I need to study this much deeper. Maybe we need to refactor the while way we access the records? Maybe any access to the records needs to be guarded with a lock, just to be safe?
If there are concurrent updates, which are guarded by lock, then should not all reads also be guarded? In the end maybe we just need to make all accesses mutually exclusive, right?

Reply:
Yeah, reading data from the extra data section, requires holding the lock.